docs/doc/60-contributing/03-rfcs/20221105-infer-schema.md

```diff
@@ -0,0 +1,160 @@
+---
+title: Infer Schema
+---
+
+- RFC PR: [datafuselabs/databend#8645](https://github.com/datafuselabs/databend/pull/8645)
+- Tracking Issue: [datafuselabs/databend#8646](https://github.com/datafuselabs/databend/issues/8646)
+
+## Summary
+
+Allowing infer schema from staged files or locations.
+
+## Motivation
+
+To load data from a stage or location, users need to create table first. However, sometimes users don't know the file schema, or the schema is too complex / too simple to be input by users.
+
+Allowing infer schema from existing files makes our users' lives easier. Also, this feature will unlock the databend from implementing `select * from @my_stage`.
+
+## Guide-level explanation
+
+Users will be able to infer the schema of the staged file by:
+
+```sql
+INFER @my_stage/data.csv FILE_FORMAT = ( TYPE = CSV);
+
++-------------+---------+----------+
+| COLUMN_NAME | TYPE    | NULLABLE |
+|-------------+---------+----------|
+| CONTINENT   | TEXT    | True     |
+| COUNTRY     | VARIANT | True     |
++-------------+---------+----------+
+```
+
+Infer from an external location is also supported:
+
+```sql
+INFER 's3://mybucket/data.csv' FILE_FORMAT = ( TYPE = CSV );
+
++-------------+---------+----------+
+| COLUMN_NAME | TYPE    | NULLABLE |
+|-------------+---------+----------|
+| CONTINENT   | TEXT    | True     |
+| COUNTRY     | VARIANT | True     |
++-------------+---------+----------+
+```
+
+`CREATE TABLE` will support `CREATE TABLE <table> BY ( <Query> )` to accept the output from `INFER`, so users can create a table with `INFRE` directly.
+
+For example:
+
+```sql
+CREATE TABLE test BY (
+    INFER 's3://mybucket/data.csv' FILE_FORMAT = ( TYPE = CSV )
+);
+```
+
+Users can also make some changes to the output of `INFER`:
+
+```sql
+CREATE TABLE test BY (
+    SELECT UPPER(COLUMN_NAME), TYPE, TRUE from (
+        INFER 's3://mybucket/data.csv' FILE_FORMAT = ( TYPE = CSV )
```

Comment:
looks like it's a table function like ` select * from  infer($location, $file_format) `

Reply:
infer is much more complex that (location, file_format).

We need to accept stage/location like in COPY, file format options and so so. I feel it's hard to implement as a function with named args.

Reply:
I think it's more like a table function.

Reply:
Is there an existing example about table function in databend?

Reply:
Not yet, but it's in ClickHouse https://clickhouse.com/docs/en/sql-reference/table-functions/s3/

Reply:
I will take this a look.
https://github.com/datafuselabs/databend/blob/0847a31d4867c20452b39f9828a65a88f2d1eb93/src/query/service/src/table_functions/table_function_factory.rs#L65-L69